From 66a4b4b2c1f510465fd4905ff591eab6b2ad6aeb Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 8 Sep 2014 13:09:07 -0400
Subject: [tools] perf/tests: Spawn child for each test

Message-id: <1410181752-25631-303-git-send-email-jolsa@redhat.com>
Patchwork-id: 91928
O-Subject: [PATCH RHEL7.1 BZ1134356 302/307] perf tests: Spawn child for each test
Bugzilla: 1134356
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Don Zickus <dzickus@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1134356
https://bugzilla.redhat.com/show_bug.cgi?id=1134356

upstream
========
commit 0d8a5faaf5a1087c7212a6f0d81920a93396414a
Author: Jiri Olsa <jolsa@kernel.org>
Date: Sat May 10 17:22:30 2014 +0200

description
===========
In upcoming tests we will setup process limits, which
might affect other tests. Spawning child for each test
to prevent this.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/tests/builtin-test.c b/tools/perf/tests/builtin-test.c
index 802e3cd..9677a5c 100644
--- a/tools/perf/tests/builtin-test.c
+++ b/tools/perf/tests/builtin-test.c
@@ -3,6 +3,8 @@
  *
  * Builtin regression testing command: ever growing number of sanity tests
  */
+#include <unistd.h>
+#include <string.h>
 #include "builtin.h"
 #include "intlist.h"
 #include "tests.h"
@@ -172,6 +174,34 @@ static bool perf_test__matches(int curr, int argc, const char *argv[])
  return false;
 }
 
+static int run_test(struct test *test)
+{
+ int status, err = -1, child = fork();
+
+ if (child < 0) {
+  pr_err("failed to fork test: %s\n", strerror(errno));
+  return -1;
+ }
+
+ if (!child) {
+  pr_debug("test child forked, pid %d\n", getpid());
+  err = test->func();
+  exit(err);
+ }
+
+ wait(&status);
+
+ if (WIFEXITED(status)) {
+  err = WEXITSTATUS(status);
+  pr_debug("test child finished with %d\n", err);
+ } else if (WIFSIGNALED(status)) {
+  err = -1;
+  pr_debug("test child interrupted\n");
+ }
+
+ return err;
+}
+
 static int __cmd_test(int argc, const char *argv[], struct intlist *skiplist)
 {
  int i = 0;
@@ -200,7 +230,7 @@ static int __cmd_test(int argc, const char *argv[], struct intlist *skiplist)
   }
 
   pr_debug("\n--- start ---\n");
-  err = tests[curr].func();
+  err = run_test(&tests[curr]);
   pr_debug("---- end ----\n%s:", tests[curr].desc);
 
   switch (err) {
-- 
1.7.1